p2p/src/test/kotlin/maru/p2p/DiscoveryTest.kt

```diff
@@ -0,0 +1,175 @@
+    private const val PORT2 = "9235"
+    private const val PORT3 = "9236"
+    private const val PORT4 = "9237"
+    private const val PORT5 = "9238"
+    private const val PORT6 = "9239"
+
+    private const val PRIVATE_KEY1: String =
+      "0x12c0b113e2b0c37388e2b484112e13f05c92c4471e3ee1dfaa368fa5045325b2"
+    private const val PRIVATE_KEY2: String =
+      "0xf3d2fffa99dc8906823866d96316492ebf7a8478713a89a58b7385af85b088a1"
+    private const val PRIVATE_KEY3: String =
+      "0x4437acb8e84bc346f7640f239da84abe99bc6f97b7855f204e34688d2977fd57"
+
+    private val key1 = Bytes.fromHexString(PRIVATE_KEY1).toArray()
+    private val key2 = Bytes.fromHexString(PRIVATE_KEY2).toArray()
+    private val key3 = Bytes.fromHexString(PRIVATE_KEY3).toArray()
+  }
+
+  @Test
+  fun `discovery finds nodes`() {
+    val bootnode =
+      DiscoveryService(
+        privateKeyBytes = key1,
+        p2pConfig =
+          P2P(
+            ipAddress = IPV4,
+            port = PORT1,
+            discoveryPort = PORT2,
+            bootnodes = emptyList(),
+          ),
+      )
+
+    val enrString = getBootnodeEnrString()
+
+    val discoveryService2 =
+      DiscoveryService(
+        privateKeyBytes = key2,
+        p2pConfig =
+          P2P(
+            ipAddress = IPV4,
+            port = PORT3,
+            discoveryPort = PORT4,
+            bootnodes = listOf(enrString),
+          ),
+      )
+
+    val discoveryService3 =
+      DiscoveryService(
+        privateKeyBytes = key3,
+        p2pConfig =
+          P2P(
+            ipAddress = IPV4,
+            port = PORT5,
+            discoveryPort = PORT6,
+            bootnodes = listOf(enrString),
+          ),
+      )
+
+    try {
+      bootnode.start()
+      discoveryService2.start()
+      discoveryService3.start()
+
+      Awaitility
+        .await()
+        .timeout(10, TimeUnit.SECONDS)
+        .untilAsserted {
+          val get = discoveryService2.searchForPeers().get()
+          assertThat(
+            get
+              .stream()
+              .filter { it.nodeId == discoveryService3.getLocalNodeRecord().nodeId }
+              .count(),
+          ).isGreaterThan(0L)
+        }
+      Awaitility
+        .await()
+        .timeout(10, TimeUnit.SECONDS)
+        .untilAsserted {
+          val get = discoveryService3.searchForPeers().get()
+          assertThat(
+            get
+              .stream()
+              .filter { it.nodeId == discoveryService2.getLocalNodeRecord().nodeId }
+              .count(),
+          ).isGreaterThan(0L)
+        }
+      Awaitility
+        .await()
+        .timeout(10, TimeUnit.SECONDS)
+        .untilAsserted {
+          val get = bootnode.searchForPeers().get()
+          assertThat(
+            get
+              .stream()
+              .filter { it.nodeId == discoveryService2.getLocalNodeRecord().nodeId }
+              .count(),
+          ).isGreaterThan(0L)
+        }
+      Awaitility
```

Comment:
I feel like this wait can be factored in a separate function for better readability and code reuse

Reply:
Done